storage/watchable_store_test.go

```diff
@@ -59,6 +58,151 @@ func TestNewWatcherCancel(t *testing.T) {
+	s.Put(testKey, testValue)
+
+	w := s.NewWatcher()
+
+	// arbitrary number for watcher
+	watcherSize := 100
+
+	// create watcherSize of CancelFunc of
+	// synced and unsynced
+	cancels := make([]CancelFunc, 2*watcherSize)
+	for i := 0; i < watcherSize; i++ {
+		// use 1 to keep watchers in unsynced
+		_, cancel := w.Watch(testKey, true, 1)
+		cancels[i] = cancel
+	}
+	for i := watcherSize; i < 2*watcherSize; i++ {
+		// use 0 to keep watchers in synced
+		_, cancel := w.Watch(testKey, true, 0)
+		cancels[i] = cancel
+	}
+
+	for idx := range cancels {
+		cancels[idx]()
+	}
+
+	// After running CancelFunc
+	//
+	// synced should be empty
+	// because cancel removes watching from synced
+	if len(s.synced[string(testKey)]) != 0 {
+		t.Errorf("synced[string(testKey)] size = %d, want 0", len(s.synced[string(testKey)]))
+	}
+	// unsynced should be empty
+	// because cancel removes watching from unsynced
+	if len(s.unsynced) != 0 {
+		t.Errorf("unsynced size = %d, want 0", len(s.unsynced))
+	}
+}
+
+// TestSyncWatchings populates unsynced watching map and
+// tests syncWatchings method to see if it correctly sends
+// events to channel of unsynced watchings and move these
+// watchings to synced.
+func TestSyncWatchings(t *testing.T) {
+	s := &watchableStore{
+		store:    newStore(tmpPath),
+		unsynced: make(map[*watching]struct{}),
+		synced:   make(map[string]map[*watching]struct{}),
+	}
+
+	defer func() {
+		s.store.Close()
+		os.Remove(tmpPath)
+	}()
+
+	testKey := []byte("foo")
+	testValue := []byte("bar")
+	s.Put(testKey, testValue)
+
+	w := s.NewWatcher()
+
+	// arbitrary number for watcher
+	watcherSize := 100
+
+	for i := 0; i < watcherSize; i++ {
+		// use 1 to keep watchers in unsynced
+		w.Watch(testKey, true, 1)
+	}
+
+	// Before running s.syncWatchings()
+	//
+	// synced should be empty
+	// because we manually populate unsynced only
+	if len(s.synced[string(testKey)]) != 0 {
+		t.Fatalf("synced[string(testKey)] size = %d, want 0", len(s.synced[string(testKey)]))
+	}
+	// unsynced should not be empty
+	// because we manually populated unsynced only
+	if len(s.unsynced) == 0 {
+		t.Errorf("unsynced size = %d, want %d", len(s.unsynced), watcherSize)
+	}
+
+	// this should move all unsynced watchings
+	// to synced ones
+	s.syncWatchings()
+
+	// After running s.syncWatchings()
+	//
+	// synced should not be empty
+	// because syncWatchings populates synced
+	// in this test case
+	if len(s.synced[string(testKey)]) == 0 {
+		t.Errorf("synced[string(testKey)] size = 0, want %d", len(s.synced[string(testKey)]))
+	}
+	// unsynced should be empty
+	// because syncWatchings is expected to move
+	// all watchings from unsynced to synced
+	// in this test case
+	if len(s.unsynced) != 0 {
+		t.Errorf("unsynced size = %d, want 0", len(s.unsynced))
+	}
+
+	// All of the watchings actually share one channel
+	// so we only need to check one shared channel
+	// (See watcher.go for more detail).
+	if len(w.(*watcher).ch) != len(s.synced)*watcherSize {
```

Comment:
We should remove `len(s.synced)`. We should expect `len(w.(*watcher).ch) == watcherSize`. 

We have N watchers and each should receive one event, which is the put at line 145.

This passes the test since len(s.synced) = 1.